packages/firestore/src/local/indexeddb_query_cache.ts

```diff
@@ -97,36 +97,75 @@ export class IndexedDbQueryCache implements QueryCache {
     );
   }
 
-  addQueryData(
+  private saveQueryData(
     transaction: PersistenceTransaction,
     queryData: QueryData
   ): PersistencePromise<void> {
-    const targetId = queryData.targetId;
-    const addedQueryPromise = targetsStore(transaction).put(
-      this.serializer.toDbTarget(queryData)
-    );
-    if (targetId > this.metadata.highestTargetId) {
-      this.metadata.highestTargetId = targetId;
-      return addedQueryPromise.next(() =>
-        globalTargetStore(transaction).put(DbTargetGlobal.key, this.metadata)
-      );
-    } else {
-      return addedQueryPromise;
+    return targetsStore(transaction).put(this.serializer.toDbTarget(queryData));
+  }
+
+  // Updates the in-memory version of the metadata. Saving is done separately.
+  // Returns true if there were any changes to the metadata.
+  private updateMetadata(queryData: QueryData): boolean {
+    let needsUpdate = false;
+    if (queryData.targetId > this.metadata.highestTargetId) {
+      this.metadata.highestTargetId = queryData.targetId;
+      needsUpdate = true;
     }
+
+    // TODO(gsoltis): add sequence number check
+    return needsUpdate;
   }
 
-  removeQueryData(
+  private saveMetadata(
+    transaction: PersistenceTransaction
+  ): PersistencePromise<void> {
+    return globalTargetStore(transaction).put(
+      DbTargetGlobal.key,
+      this.metadata
+    );
+  }
+
+  addQueryData(
+    transaction: PersistenceTransaction,
+    queryData: QueryData
+  ): PersistencePromise<void> {
```

Comment:
If it were me I'd add logic to verify that the query data doesn't already exist (asserting otherwise) as I suggested for the in-memory implementation. This is a perf hit, so I'd add a TODO to revisit for performance.

That said, having just the memory implementation enforce the contract is _probably_ sufficient, so I'm okay with leaving it out here. I'd perhaps add a comment to make the expectation clear though, e.g.:

// Although we require that the target does not already exist, we don't enforce it here for performance (though MemoryQueryCache does).

[and similar for the updateQueryData() method]

Reply:
I'd rather not add the read, as in practice we actually do the read before calling this anyways. I am intending on encapsulating this behavior though in the rethink of the interaction between `LocalStore` and the persistence layer. My current prototype calls for `getOrCreateQueryData` and various `updateQuery` methods that take a target ID and do a `get` first. The raw `addQueryData` and `updateQueryData` would no longer be exposed to the `LocalStore`.